clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RetryManager.java

```diff
@@ -0,0 +1,121 @@
+package com.linkedin.venice.fastclient.meta;
+
+import com.linkedin.venice.throttle.TokenBucket;
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class offers advanced client retry behaviors. Specifically enforcing a retry budget and relevant monitoring to
+ * avoid retry storm and alert users when the retry threshold is misconfigured or service is degrading.
+ */
+public class RetryManager implements Closeable {
+  private static final Logger LOGGER = LogManager.getLogger(RetryManager.class);
+  private static final int TOKEN_BUCKET_REFILL_INTERVAL_IN_SECONDS = 1;
+  private static final int TOKEN_BUCKET_CAPACITY_MULTIPLE = 5;
+  private final AtomicBoolean retryBudgetEnabled = new AtomicBoolean();
+  private final long enforcementWindowInMs;
+  private final double retryBudgetInPercentDecimal;
+  private final AtomicLong requestCount = new AtomicLong();
+  private final AtomicReference<TokenBucket> retryTokenBucket = new AtomicReference<>(null);
+  private final ScheduledExecutorService scheduler;
+  private final Clock clock;
+
+  private long lastUpdateTimestamp;
+  private long previousQPS = 0;
+
+  public RetryManager(long enforcementWindowInMs, double retryBudgetInPercentDecimal, Clock clock) {
+    if (enforcementWindowInMs <= 0 || retryBudgetInPercentDecimal <= 0) {
+      scheduler = null;
+      retryBudgetEnabled.set(false);
+    } else {
+      scheduler = Executors.newScheduledThreadPool(1);
+      retryBudgetEnabled.set(true);
+      lastUpdateTimestamp = clock.millis();
+      scheduler.schedule(this::updateRetryTokenBucket, enforcementWindowInMs, TimeUnit.MILLISECONDS);
+    }
+    this.enforcementWindowInMs = enforcementWindowInMs;
+    this.retryBudgetInPercentDecimal = retryBudgetInPercentDecimal;
+    this.clock = clock;
+  }
+
+  public RetryManager(long enforcementWindowInMs, double retryBudgetInPercentDecimal) {
+    this(enforcementWindowInMs, retryBudgetInPercentDecimal, Clock.systemUTC());
+  }
+
+  public void recordRequest() {
+    if (retryBudgetEnabled.get()) {
+      requestCount.incrementAndGet();
+    }
+  }
+
+  public boolean isRetryAllowed() {
+    if (!retryBudgetEnabled.get() || retryTokenBucket.get() == null) {
+      // All retries are allowed when the feature is disabled or during the very first enforcement window when we
+      // haven't collected enough data points yet
+      return true;
+    }
+    return retryTokenBucket.get().tryConsume(1);
```

Comment:
As `retryTokenBucket` can change, so shall we keep a local reference to avoid the potentially race conditions?
```
TokenBucket tokenBucket = retryTokenBucket.get();
if (!retryBudgetEnabled.get() || tokenBucket == null) {
      // All retries are allowed when the feature is disabled or during the very first enforcement window when we
      // haven't collected enough data points yet
      return true;
}
return tokenBucket.tryConsume(1);
```

Reply:
Keeping a local reference just changes the behavior but doesn't really avoid any potential race conditions IMO. We are not taking a lock and blocking when consuming and and updating `retryTokenBucket`. We could get the local reference before or after `retryTokenBucket` is updated.

Reply:
yeah, and this code I mentioned is mostly to avoid the case that first call of `retryTokenBucket.get()` is not `null`, but the second call of `retryTokenBucket.get()` is `null`.